pkg/task/restore_log.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2020 PingCAP, Inc. Licensed under Apache-2.0.
+
+package task
+
+import (
+	"context"
+
+	"github.com/pingcap/errors"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"github.com/pingcap/br/pkg/glue"
+	"github.com/pingcap/br/pkg/restore"
+	"github.com/pingcap/br/pkg/storage"
+)
+
+const (
+	flagStartTS         = "start-ts"
+	flagEndTS           = "end-ts"
+	flagBatchWriteCount = "write-kvs"
+	flagBatchFlushCount = "flush-kvs"
+
+	// represents kv size flush to storage for each table.
+	defaultFlushKV = 5120
+	// represents kv size that write to TiKV once at at time.
+	defaultWriteKV = 1280
+)
+
+// LogRestoreConfig is the configuration specific for restore tasks.
+type LogRestoreConfig struct {
+	Config
+
+	StartTS uint64
+	EndTS   uint64
+
+	BatchFlushKVPairs int
+	BatchWriteKVPairs int
+}
+
+// DefineLogRestoreFlags defines common flags for the backup command.
+func DefineLogRestoreFlags(command *cobra.Command) {
+	command.Flags().Uint64P(flagStartTS, "", 0, "restore log start ts")
+	command.Flags().Uint64P(flagEndTS, "", 0, "restore log end ts")
+
+	command.Flags().Uint64P(flagBatchWriteCount, "", 0, "the kv count that write to TiKV once at a time")
+	command.Flags().Uint64P(flagBatchFlushCount, "", 0, "the kv count that flush from memory to TiKV")
+}
+
+// ParseFromFlags parses the restore-related flags from the flag set.
+func (cfg *LogRestoreConfig) ParseFromFlags(flags *pflag.FlagSet) error {
+	var err error
+	cfg.StartTS, err = flags.GetUint64(flagStartTS)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cfg.EndTS, err = flags.GetUint64(flagEndTS)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = cfg.Config.ParseFromFlags(flags)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// adjustRestoreConfig is use for BR(binary) and BR in TiDB.
+// When new config was add and not included in parser.
+// we should set proper value in this function.
+// so that both binary and TiDB will use same default value.
+func (cfg *LogRestoreConfig) adjustRestoreConfig() {
+	if cfg.Config.Concurrency == 0 {
+		cfg.Config.Concurrency = defaultRestoreConcurrency
+	}
+	if cfg.BatchFlushKVPairs == 0 {
+		cfg.BatchFlushKVPairs = defaultFlushKV
```

Comment:
Shall we make the flush triggered both by kv pairs and total size? In the case if each kv is quite large (maybe upto sevral MB), flush only by kv pairs maybe not sufficient

Reply:
I add the flush size config to control the flush